File: frontend/src/api/purchases.ts
import { request, requestCollection } from "./client";
import { ReturnDisposition, ReturnReasonCategory } from "./types";

export type PurchaseOrderStatus =
  | "BORRADOR"
  | "PENDIENTE"
  | "APROBADA"
  | "ENVIADA"
  | "PARCIAL"
  | "COMPLETADA"
  | "CANCELADA";

export type PurchaseOrderDocument = {
  id: number;
  purchase_order_id: number;
  filename: string;
  content_type: string;
  storage_backend: string;
  uploaded_at: string;
  uploaded_by_id: number | null;
  download_url?: string | null;
};

export type PurchaseOrderStatusEvent = {
  id: number;
  purchase_order_id: number;
  status: PurchaseOrderStatus;
  note?: string | null;
  created_at: string;
  created_by_id: number | null;
  created_by_name?: string | null;
};

export type PurchaseOrderItem = {
  id: number;
  purchase_order_id: number;
  device_id: number;
  quantity_ordered: number;
  quantity_received: number;
  unit_cost: number;
};

export type PurchaseReturn = {
  id: number;
  purchase_order_id: number;
  device_id: number;
  quantity: number;
  reason: string;
  reason_category: ReturnReasonCategory;
  disposition: ReturnDisposition;
  warehouse_id?: number | null;
  supplier_ledger_entry_id?: number | null;
  corporate_reason?: string | null;
  credit_note_amount: number;
  processed_by_id: number | null;
  approved_by_id?: number | null;
  approved_by_name?: string | null;
  receipt_pdf_base64?: string | null;
  receipt_url?: string | null;
  created_at: string;
};

export type PurchaseOrder = {
  id: number;
  store_id: number;
  supplier: string;
  status: PurchaseOrderStatus;
  notes?: string | null;
  created_at: string;
  updated_at: string;
  created_by_id?: number | null;
  closed_at?: string | null;
  items: PurchaseOrderItem[];
  returns: PurchaseReturn[];
  documents: PurchaseOrderDocument[];
  status_history: PurchaseOrderStatusEvent[];
};

export type PurchaseOrderStatusUpdateInput = {
  status: PurchaseOrderStatus;
  note?: string | null;
};

export type PurchaseOrderEmailInput = {
  recipients: string[];
  message?: string | null;
  include_documents?: boolean;
};

export type PurchaseOrderCreateInput = {
  store_id: number;
  supplier: string;
  items: { device_id: number; quantity_ordered: number; unit_cost: number }[];
  notes?: string;
};

export type PurchaseReceiveInput = {
  items: { device_id: number; quantity: number; batch_code?: string | null }[];
};

export type PurchaseReturnInput = {
  device_id: number;
  quantity: number;
  reason: string;
  disposition?: ReturnDisposition;
  warehouse_id?: number | null;
  category?: ReturnReasonCategory;
};

export type PurchaseImportResponse = {
  imported: number;
  orders: PurchaseOrder[];
  errors: string[];
};

export type PurchaseSuggestionItem = {
  store_id: number;
  store_name: string;
  supplier_id: number | null;
  supplier_name: string | null;
  device_id: number;
  sku: string;
  name: string;
  current_quantity: number;
  minimum_stock: number;
  suggested_quantity: number;
  average_daily_sales: number;
  projected_coverage_days: number | null;
  last_30_days_sales: number;
  unit_cost: number;
  reason: "below_minimum" | "projected_consumption";
  suggested_value: number;
};

export type PurchaseSuggestionStore = {
  store_id: number;
  store_name: string;
  total_suggested: number;
  total_value: number;
  items: PurchaseSuggestionItem[];
};

export type PurchaseSuggestionsResponse = {
  generated_at: string;
  lookback_days: number;
  planning_horizon_days: number;
  minimum_stock: number;
  total_items: number;
  stores: PurchaseSuggestionStore[];
};

export type PurchaseVendor = {
  id_proveedor: number;
  nombre: string;
  telefono?: string | null;
  correo?: string | null;
  direccion?: string | null;
  tipo?: string | null;
  notas?: string | null;
  estado: string;
  total_compras: number;
  total_impuesto: number;
  compras_registradas: number;
  ultima_compra?: string | null;
};

export type PurchaseVendorPayload = {
  nombre: string;
  telefono?: string;
  correo?: string;
  direccion?: string;
  tipo?: string;
  notas?: string;
  estado?: string;
};

export type PurchaseVendorStatusPayload = {
  estado: string;
};

export type PurchaseRecordItem = {
  id_detalle: number;
  producto_id: number;
  cantidad: number;
  costo_unitario: number;
  subtotal: number;
  producto_nombre?: string | null;
};

export type PurchaseRecord = {
  id_compra: number;
  proveedor_id: number;
  proveedor_nombre: string;
  usuario_id: number;
  usuario_nombre?: string | null;
  fecha: string;
  forma_pago: string;
  estado: string;
  subtotal: number;
  impuesto: number;
  total: number;
  items: PurchaseRecordItem[];
};

export type PurchaseRecordItemPayload = {
  producto_id: number;
  cantidad: number;
  costo_unitario: number;
};

export type PurchaseRecordPayload = {
  proveedor_id: number;
  forma_pago: string;
  estado?: string;
  impuesto_tasa?: number;
  fecha?: string;
  items: PurchaseRecordItemPayload[];
};

export type PurchaseVendorHistory = {
  proveedor: PurchaseVendor;
  compras: PurchaseRecord[];
  total: number;
  impuesto: number;
  registros: number;
};

export type PurchaseVendorRanking = {
  vendor_id: number;
  vendor_name: string;
  total: number;
  orders: number;
};

export type PurchaseUserRanking = {
  user_id: number;
  user_name?: string | null;
  total: number;
  orders: number;
};

export type PurchaseStatistics = {
  updated_at: string;
  compras_registradas: number;
  total: number;
  impuesto: number;
  monthly_totals: { label: string; value: number }[];
  top_vendors: PurchaseVendorRanking[];
  top_users: PurchaseUserRanking[];
};

export function listPurchaseOrders(token: string, storeId: number, limit = 50): Promise<PurchaseOrder[]> {
  const params = new URLSearchParams({ limit: String(limit), store_id: String(storeId) });
  return requestCollection<PurchaseOrder>(
    `/purchases/?${params.toString()}`,
    { method: "GET" },
    token,
  );
}

export function getPurchaseOrder(token: string, orderId: number): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(`/purchases/${orderId}`, { method: "GET" }, token);
}

export function createPurchaseOrder(
  token: string,
  payload: PurchaseOrderCreateInput,
  reason: string
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    "/purchases",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function receivePurchaseOrder(
  token: string,
  orderId: number,
  payload: PurchaseReceiveInput,
  reason: string
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    `/purchases/${orderId}/receive`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function cancelPurchaseOrder(
  token: string,
  orderId: number,
  reason: string
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    `/purchases/${orderId}/cancel`,
    {
      method: "POST",
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function uploadPurchaseOrderDocument(
  token: string,
  orderId: number,
  file: File,
  reason: string
): Promise<PurchaseOrderDocument> {
  const formData = new FormData();
  formData.append("file", file);
  return request<PurchaseOrderDocument>(
    `/purchases/${orderId}/documents`,
    {
      method: "POST",
      body: formData,
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function transitionPurchaseOrderStatus(
  token: string,
  orderId: number,
  payload: PurchaseOrderStatusUpdateInput,
  reason: string
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    `/purchases/${orderId}/status`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function sendPurchaseOrderEmail(
  token: string,
  orderId: number,
  payload: PurchaseOrderEmailInput
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    `/purchases/${orderId}/send`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
    token
  );
}

export function registerPurchaseReturn(
  token: string,
  orderId: number,
  payload: PurchaseReturnInput,
  reason: string
): Promise<PurchaseReturn> {
  return request<PurchaseReturn>(
    `/purchases/${orderId}/returns`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function importPurchaseOrdersCsv(
  token: string,
  file: File,
  reason: string
): Promise<PurchaseImportResponse> {
  const formData = new FormData();
  formData.append("file", file);
  return request<PurchaseImportResponse>(
    "/purchases/import",
    {
      method: "POST",
      body: formData,
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function getPurchaseSuggestions(
  token: string,
  params: { storeId?: number; lookbackDays?: number; minimumStock?: number; planningHorizonDays?: number } = {},
): Promise<PurchaseSuggestionsResponse> {
  const query = new URLSearchParams();
  if (params.storeId != null) {
    query.set("store_id", String(params.storeId));
  }
  if (params.lookbackDays != null) {
    query.set("lookback_days", String(params.lookbackDays));
  }
  if (params.minimumStock != null) {
    query.set("minimum_stock", String(params.minimumStock));
  }
  if (params.planningHorizonDays != null) {
    query.set("planning_horizon_days", String(params.planningHorizonDays));
  }
  const suffix = query.toString();
  const path = suffix ? `/purchases/suggestions?${suffix}` : "/purchases/suggestions";
  return request<PurchaseSuggestionsResponse>(path, { method: "GET" }, token);
}

export function createPurchaseOrderFromSuggestion(
  token: string,
  payload: PurchaseOrderCreateInput,
  reason: string
): Promise<PurchaseOrder> {
  return request<PurchaseOrder>(
    "/purchases/suggestions/orders",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function listPurchaseVendors(
  token: string,
  filters: { query?: string; status?: string; limit?: number } = {}
): Promise<PurchaseVendor[]> {
  const params = new URLSearchParams();
  if (filters.query) {
    params.append("q", filters.query);
  }
  if (filters.status) {
    params.append("estado", filters.status);
  }
  if (typeof filters.limit === "number") {
    params.append("limit", String(filters.limit));
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return requestCollection<PurchaseVendor>(
    `/purchases/vendors${suffix}`,
    { method: "GET" },
    token,
  );
}

export function createPurchaseVendor(
  token: string,
  payload: PurchaseVendorPayload,
  reason: string
): Promise<PurchaseVendor> {
  return request<PurchaseVendor>(
    "/purchases/vendors",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token,
  );
}

export function updatePurchaseVendor(
  token: string,
  vendorId: number,
  payload: PurchaseVendorPayload,
  reason: string
): Promise<PurchaseVendor> {
  return request<PurchaseVendor>(
    `/purchases/vendors/${vendorId}`,
    { method: "PUT", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token,
  );
}

export function setPurchaseVendorStatus(
  token: string,
  vendorId: number,
  payload: PurchaseVendorStatusPayload,
  reason: string
): Promise<PurchaseVendor> {
  return request<PurchaseVendor>(
    `/purchases/vendors/${vendorId}/status`,
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token,
  );
}

export async function exportPurchaseVendorsCsv(
  token: string,
  filters: { query?: string; status?: string } = {},
  reason: string,
): Promise<Blob> {
  const params = new URLSearchParams();
  if (filters.query) {
    params.append("q", filters.query);
  }
  if (filters.status) {
    params.append("estado", filters.status);
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return request<Blob>(
    `/purchases/vendors/export/csv${suffix}`,
    { method: "GET", headers: { "X-Reason": reason }, responseType: "blob" },
    token,
  );
}

export function getPurchaseVendorHistory(
  token: string,
  vendorId: number,
  filters: { limit?: number; dateFrom?: string; dateTo?: string } = {}
): Promise<PurchaseVendorHistory> {
  const params = new URLSearchParams();
  if (typeof filters.limit === "number") {
    params.append("limit", String(filters.limit));
  }
  if (filters.dateFrom) {
    params.append("date_from", filters.dateFrom);
  }
  if (filters.dateTo) {
    params.append("date_to", filters.dateTo);
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return request<PurchaseVendorHistory>(
    `/purchases/vendors/${vendorId}/history${suffix}`,
    { method: "GET" },
    token,
  );
}

export function listPurchaseRecords(
  token: string,
  filters: {
    proveedorId?: number;
    usuarioId?: number;
    dateFrom?: string;
    dateTo?: string;
    estado?: string;
    query?: string;
    limit?: number;
    offset?: number;
  } = {}
): Promise<PurchaseRecord[]> {
  const params = new URLSearchParams();
  if (filters.proveedorId) {
    params.append("proveedor_id", String(filters.proveedorId));
  }
  if (filters.usuarioId) {
    params.append("usuario_id", String(filters.usuarioId));
  }
  if (filters.dateFrom) {
    params.append("date_from", filters.dateFrom);
  }
  if (filters.dateTo) {
    params.append("date_to", filters.dateTo);
  }
  if (filters.estado) {
    params.append("estado", filters.estado);
  }
  if (filters.query) {
    params.append("q", filters.query);
  }
  if (typeof filters.limit === "number") {
    params.append("limit", String(filters.limit));
  }
  if (typeof filters.offset === "number") {
    params.append("offset", String(filters.offset));
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return requestCollection<PurchaseRecord>(
    `/purchases/records${suffix}`,
    { method: "GET" },
    token,
  );
}

export function createPurchaseRecord(
  token: string,
  payload: PurchaseRecordPayload,
  reason: string
): Promise<PurchaseRecord> {
  return request<PurchaseRecord>(
    "/purchases/records",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token,
  );
}

export function exportPurchaseRecordsPdf(
  token: string,
  filters: {
    proveedorId?: number;
    usuarioId?: number;
    dateFrom?: string;
    dateTo?: string;
    estado?: string;
    query?: string;
  } = {},
  reason: string,
): Promise<Blob> {
  const params = new URLSearchParams();
  if (filters.proveedorId) {
    params.append("proveedor_id", String(filters.proveedorId));
  }
  if (filters.usuarioId) {
    params.append("usuario_id", String(filters.usuarioId));
  }
  if (filters.dateFrom) {
    params.append("date_from", filters.dateFrom);
  }
  if (filters.dateTo) {
    params.append("date_to", filters.dateTo);
  }
  if (filters.estado) {
    params.append("estado", filters.estado);
  }
  if (filters.query) {
    params.append("q", filters.query);
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return request<Blob>(
    `/purchases/records/export/pdf${suffix}`,
    { method: "GET", headers: { "X-Reason": reason }, responseType: "blob" },
    token,
  );
}

export function exportPurchaseRecordsExcel(
  token: string,
  filters: {
    proveedorId?: number;
    usuarioId?: number;
    dateFrom?: string;
    dateTo?: string;
    estado?: string;
    query?: string;
  } = {},
  reason: string,
): Promise<Blob> {
  const params = new URLSearchParams();
  if (filters.proveedorId) {
    params.append("proveedor_id", String(filters.proveedorId));
  }
  if (filters.usuarioId) {
    params.append("usuario_id", String(filters.usuarioId));
  }
  if (filters.dateFrom) {
    params.append("date_from", filters.dateFrom);
  }
  if (filters.dateTo) {
    params.append("date_to", filters.dateTo);
  }
  if (filters.estado) {
    params.append("estado", filters.estado);
  }
  if (filters.query) {
    params.append("q", filters.query);
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return request<Blob>(
    `/purchases/records/export/xlsx${suffix}`,
    { method: "GET", headers: { "X-Reason": reason }, responseType: "blob" },
    token,
  );
}

export function getPurchaseStatistics(
  token: string,
  filters: { dateFrom?: string; dateTo?: string; topLimit?: number } = {}
): Promise<PurchaseStatistics> {
  const params = new URLSearchParams();
  if (filters.dateFrom) {
    params.append("date_from", filters.dateFrom);
  }
  if (filters.dateTo) {
    params.append("date_to", filters.dateTo);
  }
  if (typeof filters.topLimit === "number") {
    params.append("top_limit", String(filters.topLimit));
  }
  const query = params.toString();
  const suffix = query ? `?${query}` : "";
  return request<PurchaseStatistics>(`/purchases/statistics${suffix}`, { method: "GET" }, token);
}

File: backend/app/schemas/purchases.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
import enum

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    computed_field,
    field_serializer,
    field_validator,
    model_validator,
)

from ..models import PurchaseStatus
from .common import DashboardChartPoint


class PurchaseOrderItemCreate(BaseModel):
    device_id: int = Field(..., ge=1)
    quantity_ordered: int = Field(..., ge=1)
    unit_cost: Decimal = Field(..., ge=Decimal("0"))

    @field_serializer("unit_cost")
    @classmethod
    def _serialize_unit_cost(cls, value: Decimal) -> float:
        return float(value)


class PurchaseOrderCreate(BaseModel):
    store_id: int = Field(..., ge=1)  # // [PACK30-31-BACKEND]
    supplier: str = Field(..., max_length=120)
    notes: str | None = Field(default=None, max_length=255)
    items: list[PurchaseOrderItemCreate]

    @model_validator(mode="before")
    @classmethod
    def _coerce_store_alias(cls, data: Any) -> Any:  # pragma: no cover - mapeo directo
        if isinstance(data, dict) and "store_id" not in data:
            for k in ("branch_id",):
                if k in data:
                    data["store_id"] = data[k]
                    break
        return data

    @field_validator("supplier")
    @classmethod
    def _validate_supplier(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Proveedor requerido")
        return normalized

    @field_validator("notes")
    @classmethod
    def _normalize_notes(cls, value: str | None) -> str | None:
        if value is None:
            return value
        normalized = value.strip()
        return normalized or None

    @field_validator("items")
    @classmethod
    def _ensure_items(cls, value: list[PurchaseOrderItemCreate]) -> list[PurchaseOrderItemCreate]:
        if not value:
            raise ValueError("Debes incluir artículos en la orden de compra.")
        return value


class PurchaseOrderItemResponse(BaseModel):
    id: int
    purchase_order_id: int
    device_id: int
    quantity_ordered: int
    quantity_received: int
    unit_cost: Decimal
    quantity_pending: int = 0

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("unit_cost")
    @classmethod
    def _serialize_unit_cost(cls, value: Decimal) -> float:
        return float(value)

    @model_validator(mode="before")
    @classmethod
    def _add_pending(cls, value: Any) -> Any:
        if isinstance(value, dict):
            quantity_ordered = int(value.get("quantity_ordered", 0) or 0)
            quantity_received = int(value.get("quantity_received", 0) or 0)
            value.setdefault("quantity_pending", max(
                quantity_ordered - quantity_received, 0))
            return value

        quantity_ordered = getattr(value, "quantity_ordered", 0) or 0
        quantity_received = getattr(value, "quantity_received", 0) or 0
        setattr(value, "quantity_pending", max(
            quantity_ordered - quantity_received, 0))
        return value


class ReturnDisposition(str, enum.Enum):
    VENDIBLE = "vendible"
    DEFECTUOSO = "defectuoso"
    NO_VENDIBLE = "no_vendible"
    REPARACION = "reparacion"


class ReturnReasonCategory(str, enum.Enum):
    DEFECTO = "defecto"
    LOGISTICA = "logistica"
    CLIENTE = "cliente"
    PRECIO = "precio"
    OTRO = "otro"


class PurchaseReturnCreate(BaseModel):
    device_id: int = Field(..., ge=1)
    quantity: int = Field(..., ge=1)
    reason: str = Field(..., min_length=5, max_length=255)
    disposition: ReturnDisposition = Field(
        default=ReturnDisposition.DEFECTUOSO
    )
    warehouse_id: int | None = Field(default=None, ge=1)
    category: ReturnReasonCategory = Field(
        default=ReturnReasonCategory.DEFECTO
    )

    @field_validator("reason")
    @classmethod
    def _normalize_reason(cls, value: str) -> str:
        normalized = value.strip()
        if len(normalized) < 5:
            raise ValueError("El motivo debe tener al menos 5 caracteres.")
        return normalized


class PurchaseReturnResponse(BaseModel):
    id: int
    purchase_order_id: int
    device_id: int
    quantity: int
    reason: str
    reason_category: ReturnReasonCategory
    disposition: ReturnDisposition
    warehouse_id: int | None
    supplier_ledger_entry_id: int | None = None
    corporate_reason: str | None = None
    credit_note_amount: Decimal = Field(default=Decimal("0"))
    processed_by_id: int | None
    approved_by_id: int | None
    approved_by_name: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("credit_note_amount")
    @classmethod
    def _serialize_credit_amount(cls, value: Decimal) -> float:
        return float(value)


class PurchaseOrderDocumentResponse(BaseModel):
    id: int
    purchase_order_id: int
    filename: str
    content_type: str
    storage_backend: str
    uploaded_at: datetime
    uploaded_by_id: int | None
    download_url: str | None = None

    model_config = ConfigDict(from_attributes=True)


class PurchaseOrderStatusEventResponse(BaseModel):
    id: int
    purchase_order_id: int
    status: PurchaseStatus
    note: str | None
    created_at: datetime
    created_by_id: int | None
    created_by_name: str | None = None

    model_config = ConfigDict(from_attributes=True)


class PurchaseOrderResponse(BaseModel):
    id: int
    store_id: int
    supplier: str
    status: PurchaseStatus
    notes: str | None
    created_at: datetime
    updated_at: datetime
    created_by_id: int | None
    approved_by_id: int | None = None
    approved_by_name: str | None = None
    requires_approval: bool = False
    closed_at: datetime | None
    items: list[PurchaseOrderItemResponse]
    pending_items: int = 0
    returns: list[PurchaseReturnResponse] = Field(default_factory=list)
    documents: list[PurchaseOrderDocumentResponse] = Field(
        default_factory=list)
    status_history: list[PurchaseOrderStatusEventResponse] = Field(
        default_factory=list, validation_alias="status_events"
    )

    model_config = ConfigDict(from_attributes=True)

    @model_validator(mode="before")
    @classmethod
    def _compute_pending_items(cls, value: Any) -> Any:
        data = dict(value) if isinstance(value, dict) else value
        items = data.get("items") if isinstance(
            data, dict) else getattr(data, "items", [])
        pending_total = 0
        for item in items or []:
            quantity_ordered = item.get("quantity_ordered") if isinstance(
                item, dict) else getattr(item, "quantity_ordered", 0)
            quantity_received = item.get("quantity_received") if isinstance(
                item, dict) else getattr(item, "quantity_received", 0)
            pending_total += max(int(quantity_ordered or 0) -
                                 int(quantity_received or 0), 0)

        if isinstance(data, dict):
            data.setdefault("pending_items", pending_total)
            return data

        setattr(data, "pending_items", pending_total)
        return data


class PurchaseOrderStatusUpdateRequest(BaseModel):
    status: PurchaseStatus
    note: str | None = Field(default=None, max_length=255)


class PurchaseOrderEmailRequest(BaseModel):
    recipients: list[str] = Field(..., min_length=1)
    message: str | None = Field(default=None, max_length=500)
    include_documents: bool = False


class PurchaseReceiveItem(BaseModel):
    device_id: int = Field(..., ge=1)
    quantity: int = Field(..., ge=1)
    batch_code: str | None = Field(default=None, max_length=80)

    @field_validator("batch_code")
    @classmethod
    def _normalize_batch_code(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class PurchaseReceiveRequest(BaseModel):
    items: list[PurchaseReceiveItem]

    @field_validator("items")
    @classmethod
    def _ensure_items(cls, value: list[PurchaseReceiveItem]) -> list[PurchaseReceiveItem]:
        if not value:
            raise ValueError("Debes indicar artículos a recibir.")
        return value


class PurchaseImportResponse(BaseModel):
    imported: int = Field(default=0, ge=0)
    orders: list[PurchaseOrderResponse]
    errors: list[str] = Field(default_factory=list)


class PurchaseSuggestionItem(BaseModel):
    store_id: int
    store_name: str
    supplier_id: int | None
    supplier_name: str | None
    device_id: int
    sku: str
    name: str
    current_quantity: int
    minimum_stock: int
    suggested_quantity: int
    average_daily_sales: float
    projected_coverage_days: int | None
    last_30_days_sales: int
    unit_cost: Decimal = Field(default=Decimal("0"))
    reason: Literal["below_minimum", "projected_consumption"]

    @field_serializer("unit_cost")
    @classmethod
    def _serialize_unit_cost(cls, value: Decimal) -> float:
        return float(value)

    @computed_field(return_type=float)  # type: ignore[misc]
    def suggested_value(self) -> float:
        return float(self.unit_cost * Decimal(self.suggested_quantity))


class PurchaseSuggestionStore(BaseModel):
    store_id: int
    store_name: str
    total_suggested: int
    total_value: float
    items: list[PurchaseSuggestionItem]


class PurchaseSuggestionsResponse(BaseModel):
    generated_at: datetime
    lookback_days: int
    planning_horizon_days: int
    minimum_stock: int
    total_items: int
    stores: list[PurchaseSuggestionStore]


class PurchaseVendorBase(BaseModel):
    nombre: str = Field(..., min_length=3, max_length=150)
    telefono: str | None = Field(default=None, max_length=40)
    correo: str | None = Field(default=None, max_length=120)
    direccion: str | None = Field(default=None, max_length=255)
    tipo: str | None = Field(default=None, max_length=60)
    notas: str | None = Field(default=None, max_length=255)

    @field_validator("nombre")
    @classmethod
    def _normalize_nombre(cls, value: str) -> str:
        normalized = value.strip()
        if len(normalized) < 3:
            raise ValueError("El nombre del proveedor es obligatorio.")
        return normalized

    @field_validator("telefono", "correo", "direccion", "tipo", "notas")
    @classmethod
    def _normalize_optional(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class PurchaseVendorCreate(PurchaseVendorBase):
    estado: str = Field(default="activo", max_length=40)


class PurchaseVendorUpdate(BaseModel):
    nombre: str | None = Field(default=None, min_length=3, max_length=150)
    telefono: str | None = Field(default=None, max_length=40)
    correo: str | None = Field(default=None, max_length=120)
    direccion: str | None = Field(default=None, max_length=255)
    tipo: str | None = Field(default=None, max_length=60)
    notas: str | None = Field(default=None, max_length=255)
    estado: str | None = Field(default=None, max_length=40)

    @field_validator("nombre", "telefono", "correo", "direccion", "tipo", "notas", "estado")
    @classmethod
    def _normalize_optional(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class PurchaseVendorResponse(PurchaseVendorBase):
    id: int = Field(alias="id_proveedor")
    estado: str
    total_compras: Decimal = Field(default=Decimal("0"))
    total_impuesto: Decimal = Field(default=Decimal("0"))
    compras_registradas: int = Field(default=0, ge=0)
    ultima_compra: datetime | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    @field_serializer("total_compras", "total_impuesto")
    @classmethod
    def _serialize_decimal(cls, value: Decimal) -> float:
        return float(value)


class PurchaseVendorStatusUpdate(BaseModel):
    estado: Literal["activo", "inactivo"]


class PurchaseRecordItemBase(BaseModel):
    producto_id: int = Field(..., ge=1)
    cantidad: int = Field(..., ge=1)
    costo_unitario: Decimal = Field(..., ge=Decimal("0"))

    @field_serializer("costo_unitario")
    @classmethod
    def _serialize_cost(cls, value: Decimal) -> float:
        return float(value)


class PurchaseRecordItemCreate(PurchaseRecordItemBase):
    """Detalle de ítems utilizados al registrar una compra."""


class PurchaseRecordItemResponse(PurchaseRecordItemBase):
    id: int = Field(alias="id_detalle")
    subtotal: Decimal = Field(default=Decimal("0"))
    producto_nombre: str | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    @field_serializer("subtotal")
    @classmethod
    def _serialize_subtotal(cls, value: Decimal) -> float:
        return float(value)


class PurchaseRecordCreate(BaseModel):
    proveedor_id: int = Field(..., ge=1)
    fecha: datetime | None = None
    forma_pago: str = Field(..., max_length=60)
    estado: str = Field(default="REGISTRADA", max_length=40)
    impuesto_tasa: Decimal = Field(default=Decimal(
        "0.16"), ge=Decimal("0"), le=Decimal("1"))
    items: list[PurchaseRecordItemCreate]

    @field_validator("items")
    @classmethod
    def _ensure_items(cls, value: list[PurchaseRecordItemCreate]) -> list[PurchaseRecordItemCreate]:
        if not value:
            raise ValueError("Debes agregar productos a la compra.")
        return value


class PurchaseRecordResponse(BaseModel):
    id: int = Field(alias="id_compra")
    proveedor_id: int
    proveedor_nombre: str
    usuario_id: int
    usuario_nombre: str | None = None
    fecha: datetime
    forma_pago: str
    estado: str
    subtotal: Decimal
    impuesto: Decimal
    total: Decimal
    items: list[PurchaseRecordItemResponse]

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    @field_serializer("subtotal", "impuesto", "total")
    @classmethod
    def _serialize_amount(cls, value: Decimal) -> float:
        return float(value)


class PurchaseVendorHistory(BaseModel):
    proveedor: PurchaseVendorResponse
    compras: list[PurchaseRecordResponse]
    total: Decimal
    impuesto: Decimal
    registros: int

    @field_serializer("total", "impuesto")
    @classmethod
    def _serialize_totals(cls, value: Decimal) -> float:
        return float(value)


class PurchaseReportFilters(BaseModel):
    proveedor_id: int | None = None
    usuario_id: int | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    estado: str | None = None
    query: str | None = None


class PurchaseReportTotals(BaseModel):
    count: int = Field(default=0, ge=0)
    subtotal: Decimal = Field(default=Decimal("0"))
    impuesto: Decimal = Field(default=Decimal("0"))
    total: Decimal = Field(default=Decimal("0"))

    @field_serializer("subtotal", "impuesto", "total")
    @classmethod
    def _serialize_decimal(cls, value: Decimal) -> float:
        return float(value)


class PurchaseReportItem(BaseModel):
    compra_id: int
    folio: str
    proveedor_nombre: str
    usuario_nombre: str | None
    forma_pago: str
    estado: str
    subtotal: Decimal
    impuesto: Decimal
    total: Decimal
    fecha: datetime
    items: list[PurchaseRecordItemResponse]

    @field_serializer("subtotal", "impuesto", "total")
    @classmethod
    def _serialize_decimal(cls, value: Decimal) -> float:
        return float(value)


class PurchaseReport(BaseModel):
    generated_at: datetime
    filters: PurchaseReportFilters
    totals: PurchaseReportTotals
    daily_stats: list[DashboardChartPoint]
    items: list[PurchaseReportItem]


class PurchaseVendorRanking(BaseModel):
    vendor_id: int
    vendor_name: str
    total: Decimal
    orders: int

    @field_serializer("total")
    @classmethod
    def _serialize_total(cls, value: Decimal) -> float:
        return float(value)


class PurchaseUserRanking(BaseModel):
    user_id: int
    user_name: str | None
    total: Decimal
    orders: int

    @field_serializer("total")
    @classmethod
    def _serialize_total(cls, value: Decimal) -> float:
        return float(value)


class PurchaseStatistics(BaseModel):
    monthly_total: Decimal
    monthly_count: int
    top_vendors: list[PurchaseVendorRanking]
    top_users: list[PurchaseUserRanking]
    daily_average: Decimal

    @field_serializer("monthly_total", "daily_average")
    @classmethod
    def _serialize_totals(cls, value: Decimal) -> float:
        return float(value)
